Shaderc-rs & Shaderc-sys split

knappador opened the pull request:
--snip-- #44  has almost all of the discussion leading up to the merge commit notes

fixes #44

knappador commented:
I signed it!
@googlebot

knappador committed:
Introducing shaderc-sys

Version 0.4.0 has been skipped and tag will not be used.  History for a prospective split at 0.4.0 into two repos has been obliterated.

Internal renaming in shaderc-rs/src/lib.rs ffi -> scs

The crates were split as follows:

Sys crate files
- Build dependencies
- build.rs
- ffi.rs
Rs crate files
- lib.rs
Top level
- contrib agreement
- .gitmodules
- .gitignore
- workspace Cargo.toml
- CI scripts
- license
- readme

knappador committed:
- rustfmt

knappador committed:
Re-introducing --no-default-features behavior

- `check-passthrough` must be enabled on the sys dependency to even allow setting the old `--no-defaults` behavior
- `dont-use-deprecated` feature of the sys package allows suppression of old `--no-defaults` to be set as a default in the rs package
- `dont-use-deprecated` must be **disabled** by passing `--no-defaults` to rs package to enable the old `--no-defaults` behavior in sys

The old use of --no-default-features applies to shaderc-rs, but now it needs to be passed through to shaderc-sys.  This is done by passing an inverted flag to suppress the behavior by default from shaderc-rs.  When the suppression is removed with --no-default-features, the old behavior is re-enabled and shaderc-sys looks for a libshaderc_combined.a in the out directory

When building shaderc-sys directly, only check for the passthrough flag if it could be enabled.  It's a package opt-in deprecated behavior.  It definitely needs deprecation.

antiagainst commented on shaderc-rs/src/lib.rs:
```diff
@@ -64,15 +64,15 @@ extern crate libc;
-use std::ffi::{CStr, CString};
-use std::cell::RefCell;
+use libc::{c_char, c_int, c_void, int32_t, size_t, uint32_t, uint8_t};
+use shaderc_sys as scs;
```
Nit: Can we put this line before `use libc::`? It is introducing an alias and different from others. :)

knappador replied:
I believe this got fixed, but was not marked outdated somehow

https://github.com/google/shaderc-rs/blob/f03ba636d9fa79b26c636269c0da01526a1e5ffa/shaderc-rs/src/lib.rs#L66-L79

knappador committed:
SHADERC_STATIC="./here/"

If you are here because your windows or macos builds are breaking, I apologize.  The changes were not developed on these platforms.  You will have to research older versions of the crate to establish whether they were in a "just works" state and if not, research your relevant platform options and make it work.

This commit implements several new build behaviors:

- Building libshaderc from source is now fallback behavior on Linux
- Linux will check "/usr/lib/" for libraries by default
- SPIRV and glslang libraries _may_ be picked up correctly if they exist in the same location as libshaderc, but funky combinations could break
- Specification of a static library via SHADERC_STATIC, SHADERC_LIB_DIR, or SHADERC_LIB_PATH instead of the hacky --no-default-options
- tests will build correctly when using hacky --no-default-options paths (stdc was not being included)

knappador commented:
@antiagainst  I'm not sure what's up on the Appveyor results.  I have little visibility outside of Arch Linux so far.

This PR should read fine as a series of independent commits.

[Re-implemented the `--no-defaults` behavior](https://github.com/e-nguyen/shaderc-rs/pull/1)
[Added more robust build options & behavior](https://github.com/e-nguyen/shaderc-rs/pull/2)

I primarily used [this heavily referenced guide to writing sys crates](https://kornel.ski/rust-sys-crate) to select what traditions to try to follow.  The one area I'm not yet 100% clear on is [overriding the static library location using a downstream project's `Cargo.toml`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key)

To quickly run through a thorough test of the Linux build paths on Arch linux:

```
cargo test
cargo clean
cargo test --features build-from-source
cp ../target/debug/../libshaderc_combined.a .
cargo clean
cargo test SHADERC_STATIC=./libshaderc_combined.a
cargo clean
cargo test --features check-passthrough  # will cause trigger old --no-defaults behavior on sys crate
# cp the library into the specified location
# re-run with --features check-passthrough
cargo clean
mkdir libs
cp /usr/lib/libshaderc_combined.a libs/
cp /usr/lib/libglslang.a libs/
cp /usr/lib/libSPIRV.a libs/
cp /usr/lib/libSPIRV-Tools.a libs/
cargo test SHADERC_STATIC=./libs/
```
Happy to punch some commits into working on Appveyor but I'll need some information.

knappador commented:
Handling the merge issue.  Tests for shaderc-rs fail due to linking.  I'm incorporating 367a4ddbaaacf9590a68102db62aa0a2bc3e8953 as well, which won't apply at all.

Might punch some more history later to convert the delete to a move if possible.

knappador commented:
Full test sequence fine on my machine, on Linux.  There was a regression in build.rs.  glslang being included statically broke compiling the tests.  I have no idea how I found that when I wrote down the critical hunk in #44

knappador committed:
Merge branch 'master' of https://github.com/google/shaderc-rs into default-to-system-on-linux

No actual conflicts.

knappador committed:
Manual reapplication of dfaefb0de8a06f637186f37007924ace2c29daba

Original PR: https://github.com/google/shaderc-rs/pull/47

To merge master

knappador commented:
@antiagainst   looks like all three platforms have some level of support based on tests?  Time to party yet?  :package: :tada: 

- [ ] Let's remember to toss in a version bump right before merge.  I think "0.5" is what we settled on for #44 

Feel free to keep pinging me on build issues moving forward.  

I'm nominating this for merge and not blocked on anything.

knappador committed:
build-from-source pass-through feature for shaderc-sys

knappador commented:
I'm still thinking about necessary ways to pass through the new options to shaderc-sys.  The command line settings "just work."  I'm going to add in a patch to pass through the build-from-source feature to shaderc-rs since it's just easier than figuring out a nested dependency configuration for most people.

knappador commented on shaderc-sys/src/lib.rs:
```diff
@@ -13,7 +13,8 @@
 #![allow(non_camel_case_types)]
 
-use libc::{c_char, c_int, c_void, size_t, int32_t, uint32_t};
+extern crate libc;
```
Just remembered this.  I don't have a clear explanation of why this was needed, just that it broke without the change.

antiagainst replied:
Because right now this file is moved into a new crate. Previously it is in the shaderc-rs crate, which had this in `lib.rs`.

antiagainst commented on README.md:
```diff
@@ -80,7 +73,19 @@ assert!(text_result.as_text().starts_with("; SPIR-V\n"));
 -----
 
-To build the shaderc-rs crate, the following tools must be installed and available on `PATH`:
+Building the underlying libshaderc from source takes a while.  You can provide it or use system libraries by setting the `SHADERC_STATIC` path.  On Linux, the build script also checks for independent libglslang and libSPIRV and includes them only if found.
```
Nit: Can we wrap these extra long lines into 80 characters?

antiagainst replied:
I think the explanation here is not very clear. It would be nice to document features, env vars, and lib resolution orders more clearly, like:

* The default behavior is to search and link system libraries for shaderc under the directory given by env var .... This first tries static library, then dynamic library. If none is found, then ...
* If `--no-default-features` is given, then ... This is deprecated ...
* If `build-from-source` is given, then ...

antiagainst commented on shaderc-rs/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "shaderc"
+version = "0.4.1"
```
Yep, let's bump this to 0.5.

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+    emit_stdc_link();
+}
+
+fn emit_stdc_link() {
```
Nit: I think it's more of `emit_std_cpp_link_lib` here.

antiagainst commented on shaderc-rs/src/lib.rs:
```diff
@@ -1732,13 +1769,15 @@ void main() { my_ssbo.x = 1.0; }";
-            Some(&options),
-        ).unwrap()
+        let result = c
+            .compile_into_spirv_assembly(
```
Normally it would be nice to separate these formatting changes with others. But I won't bother this time to avoid incurring more work for you. :)

antiagainst commented on shaderc-sys/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+version = "0.4.1"
+authors = ["Lei Zhang <antiagainst@gmail.com>", "Mr. Knapp <knappador@gmail.com>"]
+build = "build/build.rs"
+description = "Raw bindings for shaderc"
```
What about "Shaderc system library FFI, building, and linking" here? "Raw bindings" is not quite distinguishable with "Rust bindings" in the above `Cargo.toml`.

antiagainst commented on shaderc-sys/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+name = "shaderc_sys"
+
+[features]
+check-passthrough = []
```
Sorry I don't fully understand here. What's the purpose of `check-passthrough`? It seems we can merge it into `dont-use-deprecated` to be just one feature?

knappador replied:
I agree the name choices are not much use.  

As for the critical decision, using two features, it began with making shaderc-sys independent of shaderc-rs. 

Fundamentally, we are trying to re-export a feature of shaderc-sys's build options that begins by disabling the default features on shaderc-rs.  The only way to accomplish this is by inverting the feature.  This inversion also means any other package that doesn't set the feature would get `--no-defaults-behavior`.  Therefore, I protected this inverted feature with a second feature.

antiagainst replied:
Okay, I see, thanks!

antiagainst replied:
I think another way is just to eliminate the `--no-defaults-behavior` support for shaderc-rs entirely. It was not a satisfying solution; and now we have `SHADERC_LIB_DIR` now and it is better. WDYT?

antiagainst commented on shaderc-sys/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+
+[features]
+check-passthrough = []
+dont-use-deprecated = []
```
Can we give this a more descriptive name? `dont-use-deprecated` is quite general and it is applicable to all kinds of deprecated features. What about "use-sys-libs" or something similar?

antiagainst commented on shaderc-rs/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+
+[features]
+default = ["shaderc-sys-passthrough"] # disabling defaults inverts to use deprecated
+shaderc-sys-passthrough = ["shaderc-sys/dont-use-deprecated"]
```
What about "prefer-sys-libs" or something similar to be more descriptive?

antiagainst commented on shaderc-rs/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+assert_matches = "1.3"
+
+[features]
+default = ["shaderc-sys-passthrough"] # disabling defaults inverts to use deprecated
```
I'd suggest to make the comment here to be more clear:

```
# Disabling defaults to the deprecated behavior: not building from source,
not using system libraries, only relying on the developer to provide native
shaderc libraries
```

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+}
+
+fn main() {
+    // Deprecated --no-defaults path inhereted from before shaderc-rs & shaderc-sys split
```
Should we have a check to error out if both `CARGO_FEATURE_DONT_USE_DEPRECATED` and `CARGO_FEATURE_BUILD_FROM_SOURCE` is true?

knappador replied:
A lot of the options are indeed contradictory.  Stating the order of preference in the docs is sufficient IMO.

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+        && env::var("CARGO_FEATURE_DONT_USE_DEPRECATED").is_err()
+    {
+        let out_dir = env::var("OUT_DIR").unwrap();
+        println!("cargo:warning=USE OF --no-defaults IS DEPRECATED BEHAVIOR.");
```
`--no-deafults` is for shaderc-rs. Here we are in shaderc-sys now. :) What about just saying `USE OF 'dont-use-deprecated' IS ...` ?

knappador replied:
(could be old question) Let's back up.  The way I preserved the old behavior is so that in `shaderc-rs` you can still say:
```$ cargo test --no-default-features``` 
And it just works like it does in master.

This is done by adding the two goofy features.  I don't anticipate that anyone will explicitly do something like:
```toml
[dependencies]
shaderc-sys = {version = "0.5.0", features = ["enable-inverted-no-defaults"]
```

Basically it's definitely a shaderc-rs choice, part-and-parcel with using shaderc-rs, but has to be implemented in shaderc-sys do to the way `--no-default-features` behaves

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+    {
+        let out_dir = env::var("OUT_DIR").unwrap();
+        println!("cargo:warning=USE OF --no-defaults IS DEPRECATED BEHAVIOR.");
+        println!("cargo:warning=Requested to skip building native C++ shaderc.");
```
`Requested to skip finding system libraries or building native C++ shaderc` ?

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+
+    // Explicit overrides first
+    let mut search_dir = if let Ok(lib_dir) = env::var("SHADERC_LIB_DIR") {
+        println!("cargo:warning=Using specified pre-built libshaderc");
```
Can we make this message a little bit more explicit by attaching the dir to it: `Using specified pre-built libshaderc in {}` ? It could be quite useful if something goes wrong regarding the directory so that the developer can easily notice it. :)

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+    } else if let Ok(lib_path) = env::var("SHADERC_LIB_PATH") {
+        println!("cargo:warning=Using specified pre-built libshaderc");
+        Some(lib_path)
+    } else if let Ok(lib_path) = env::var("SHADERC_STATIC") {
```
Question: Why are we having three environment variables for the same functionality here?

knappador replied:
I'm going to change it to only use `SHADERC_LIB_DIR` since we might have multiple libraries and calling it a path would be false.

A lot of the choices attempted to defer to this document: https://kornel.ski/rust-sys-crate

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+            if combined_lib_path.exists() {
+                Some((search_dir.clone(), COMBINED_LIB.to_owned()))
+            } else if dylib_path.exists() {
+                Some((search_dir.clone(), dylib_name))
```
Should we prefer dynamic lib to static lib? What do you think? But it's fine to do in another PR.

knappador replied:
My vantage point is from years of only maintaining Linux machines, so it's actually quite hard for me to anticipate or reason about their choices.

I cargo-culted the bits that try to build a Windows dll path from the same document I linked before.  I truly have no idea what kind of library might get built on Windows.

knappador replied:
You might have been asking about dynamic on Linux.  I think I read that static was more likely to succeed.  Not specifying would be cool since it will search more accurately.

The pkg-config file has landed, but it will be a little bit before it hits package managers.  As for from source builds, I believe static is default when you bring your own version because you can only share if everyone is on the same page (using a system package manager) 
https://github.com/google/shaderc/commit/439a848c0c0e91b5f1b85059cd4b059274b9a93b

antiagainst replied:
SG! I don't have strong opinions here. As you said, it's hard to have a satisfying default anyway. :)

antiagainst commented on shaderc-sys/build/build.rs:
```diff
@@ -0,0 +1,194 @@
+
+        if let Some((lib_dir, lib_name)) = {
+            if combined_lib_path.exists() {
+                Some((search_dir.clone(), COMBINED_LIB.to_owned()))
```
I think we can do `Some((&search_dir, COMBINED_LIB))` and `Some((&search_dir, dylib_name.as_str()))` to avoid copying?

antiagainst commented on README.md:
```diff
@@ -80,7 +73,19 @@ assert!(text_result.as_text().starts_with("; SPIR-V\n"));
-To build the shaderc-rs crate, the following tools must be installed and available on `PATH`:
+Building the underlying libshaderc from source takes a while.  You can provide it or use system libraries by setting the `SHADERC_STATIC` path.  On Linux, the build script also checks for independent libglslang and libSPIRV and includes them only if found.
+
+`--no-default-features` still works on shaderc-rs, but shaderc-sys implements this behavior in a deprecated manner.  Prefer `SHADERC_STATIC="../path/to/libshaderc/and/glsang/etc/"`.  This method only works with a monolithic **libshaderc_combined.a**.
```
Could you also update the documentation in `lib.rs`? Thanks!

antiagainst reviewed (commented):
This is awesome progress! :) I've a few comments inlined. Thanks for the work again!

knappador committed:
Assorted cleanups to names and documents.  These changes are locally limited and should not be interpreted as a larger coherent change.

knappador commented:
I didn't have a lot of time to review changes in https://github.com/google/shaderc-rs/pull/46/commits/f03ba636d9fa79b26c636269c0da01526a1e5ffa

Many of the comments were addressed.  I'll punch the revision history after changes are more final.

Worth considering how the new structure might affect the docs.  I have little experience yet working with Rust docs.
https://docs.rs/shaderc/0.3.16/shaderc/

antiagainst commented on shaderc-sys/Cargo.toml:
```diff
@@ -15,9 +15,9 @@ links = "shaderc"
-dont-use-deprecated = []
 build-from-source = []
+check-inverted-no-defaults = []
+inverted-no-defaults = []
```
How do you think about `shaderc-rs-use-sys-libs` and `use-sys-libs` here then? Given the tight relationship between shaderc-rs and shaderc-sys, it is reasonable to have an option dedicated to shaderc-rs here. (I'm just trying to make the feature more self-documenting. :) )

knappador replied:
This crazy inversion thing is only necessary because of how `--no-default-features` operates on a crate's config.  I would support a rename to `shaderc` specific if you keep the backwards compatibility.  If you axe the backwards compatibility, these features go away.  `shaderc-rs-inverted-no-defaults` etc would be more appropriate.  The features have no control over reading specific directories, only the old behavior of reading the `../out/` directory.

If you decide to axe no defaults, it's as simple as removing these two features in both crates and axing the block in build.rs that reads them, then removing the documents.

The reason I implemented the `--no-default-features` backwards compatibility is because you believed people were using it (possibly in CI?).  It's wise to maintain at least one release of deprecation in any case.

antiagainst replied:
SG. This will be removed soon, so I won't bother that much anymore. :)

antiagainst reviewed (approved):
This is nice, thanks again for doing this! :)

Feel free to do whatever clean up you'd like to and then let me know. I will apply a few cosmetic fixes and then merge this.

knappador committed:
Relative paths are kind of a pain in build scripts since they are relative to the dependency, not the location of where cargo was invoked.  Warn the user if a relative path is relative, not a directory, or fails to resolve (usually file not found becuase of a bad relative path)

knappador commented:
I added some cleanups after doing some testing and sitting back to leisurely read the code.  One feature I actually changed was to try to canonicalize relative paths so that they would work... only to realize that the CWD changes during each recursion cargo makes to a dependency.  Specifying a relative path must be done relative to `shaderc-sys` crate. 

--rerun-after-changed and it's env variable counterpart deserve a look.  I think we have to combine both for our case, one to detect lib file changes and the other to detect env changes.  They have some funky behavior though.  I didn't try them.

 AFAIK the only outstanding question is to completely axe `--no-default-features` behavior or leave it deprecated for a few release cycles.


Tests
```shell
cargo test
cargo clean
cargo test --features build-from-source
cp ../target/debug/../libshaderc_combined.a ../
cargo clean
cargo test SHADERC_LIB_DIR=../
cargo clean
cargo test --no-default-features
# cp the library into the specified location
# re-run  test --no-default-features
cargo clean
mkdir ../libs
cp /usr/lib/libshaderc_combined.a ../libs/
cp /usr/lib/libglslang.a ../libs/
cp /usr/lib/libSPIRV.a ../libs/
cp /usr/lib/libSPIRV-Tools.a ../libs/
cargo test SHADERC_LIB_DIR=../libs/
```

knappador committed:
fix README path used to make docs & crates.io page

knappador committed:
Also use shaderc-rs README for crates info.  I don't have a strong opinion here since this is a combined crate's README that contains quite a bit of relevant build information for the crate in spite of a lot of shaderc-rs specific information.  Two readmes might be a bad idea for maintenance.

antiagainst committed:
Refine documentation

antiagainst commented:
Looks good! I just did some final touch over documentation. Will merge soon.

Good point about `--no-default-features` support. I think with semver, it should be fine to axe it entirely. (Sorry I didn't anticipate it as being this complicated, so I suggested we should support it earlier. We have users of it, see #36) But you already implemented it, let's be nice to keep it for another release. Thanks! :)

knappador commented:
:+1:   Changes look good.  Thanks for the extensive review, feedback, and refinements

https://github.com/vulkano-rs/vulkano/issues/1183 crosslinking

antiagainst commented:
Merged, thanks! @knappador, it seems I cannot publish shaderc-sys because I don't have the right access. Would you mind add me? Or If you want to do it, please feel free (use the top of the tree). :)

knappador commented:
I believe I invited you already, I will resend when I get to my credentials

On Sun, Apr 21, 2019, 10:16 PM Lei Zhang <notifications@github.com> wrote:

> Merged, thanks! @knappador <https://github.com/knappador>, it seems I
> cannot publish shaderc-sys because I don't have the right access. Would you
> mind add me? Or If you want to do it, please feel free (use the top of the
> tree). :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/shaderc-rs/pull/46#issuecomment-485312308>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE4Q4YGOROWGG2IAKKNQWLPRUU2PANCNFSM4HCOEWHQ>
> .
>

antiagainst commented:
https://crates.io/me/pending-invites shows I've no pending owner invites.

knappador commented:
@antiagainst 

Invite sent.  If there's no crate, the UI presents no option to manage owners :speak_no_evil: 

`cargo publish` from within the shaderc-sys crate failed first time.  I changed the README path back to relative, success.  I can PR this if you're unsure of what I did to get cargo to succeed.  

Also it means I burned another semver to give you an invite :speak_no_evil:  0.4.2 I guess :cat:

antiagainst commented:
Nice, I've access now and published 0.5.0! \o/ Thanks!

Could you yank 0.4.1 now? :)

knappador commented:
Yanked.  Going to merge this downstream for CI justice.
